packages/fxa-auth-server/lib/routes/utils/subscriptions.js

```diff
@@ -29,43 +29,131 @@ module.exports = {
-        capabilities.concat(productCapabilities[product] || []),
-      []
-    );
-
-    const clientVisibleCapabilities = clientCapabilities[client_id] || [];
-
-    const capabilitiesToReveal = new Set(
-      subscribedCapabilities.filter(
-        capability =>
-          auth.strategy === 'sessionToken' ||
-          clientVisibleCapabilities.includes(capability)
-      )
+    // TODO: issue #3913 - remove support for capabilities from server config
+    checkCapabilitiesFromServerConfig(
+      capabilitiesToReveal,
+      subscribedProducts,
+      client_id,
+      config,
+      auth
     );
 
     return capabilitiesToReveal.size > 0
       ? Array.from(capabilitiesToReveal)
       : undefined;
   },
-};
+});
+
+async function fetchSubscribedProductsFromStripe(
+  subscribedProducts,
+  uid,
+  stripeHelper,
+  email
+) {
+  const customer = await stripeHelper.customer(uid, email);
+  if (!customer || !customer.subscriptions.data) {
+    return;
+  }
+  subscribedProducts.push(
+    ...customer.subscriptions.data.map(
+      ({ plan: { product: productId } }) => productId
+    )
+  );
+}
+
+async function checkCapabilitiesFromStripe(
+  subscribedProducts,
+  capabilitiesToReveal,
+  client_id,
+  stripeHelper
+) {
+  // Run through all plans and collect capabilitiies for subscribed products
+  const plans = await stripeHelper.allPlans();
+  for (const plan of plans) {
+    if (!subscribedProducts.includes(plan.product_id)) {
+      continue;
+    }
+    const metadata = SubscriptionUtils.metadataFromPlan(plan);
+    const capabilityKeys = [
+      'capabilities',
+      ...Object.keys(metadata).filter(key =>
+        client_id === null
+          ? key.startsWith('capabilities:')
+          : key === `capabilities:${client_id}`
+      ),
+    ].filter(key => key in metadata);
+    for (const key of capabilityKeys) {
+      const capabilities = metadata[key].split(',');
+      for (const capability of capabilities) {
+        capabilitiesToReveal.add(capability);
+      }
+    }
+  }
+}
+
+// TODO: issue #3846 - remove this method
+async function fetchSubscribedProductsFromLocalDB(subscribedProducts, db, uid) {
+  const subscriptions = (await db.fetchAccountSubscriptions(uid)) || [];
+  subscribedProducts.push(...subscriptions.map(({ productId }) => productId));
+}
+
+// TODO: issue #3913 - remove support for capabilities from server config
+function checkCapabilitiesFromServerConfig(
+  capabilitiesToReveal,
+  subscribedProducts,
+  client_id,
+  config,
+  auth
+) {
+  const {
+    subscriptions: { productCapabilities = {}, clientCapabilities = {} } = {},
+  } = config;
+
+  const subscribedCapabilities = subscribedProducts.reduce(
+    (capabilities, product) =>
+      capabilities.concat(productCapabilities[product] || []),
+    []
+  );
+
+  const clientVisibleCapabilities = clientCapabilities[client_id] || [];
+
+  subscribedCapabilities
+    .filter(
+      capability =>
+        auth.strategy === 'sessionToken' ||
```

Comment:
I don't understand why all the capabilities are returned in this case, I'm probably missing some context.

Reply:
We might be able to drop this, but at one point IIRC we wanted to have all the capabilities accessible to content-server using sessionToken